[Fix #7737] Add new `Style/UnnecessaryArguments` cop

tejasbubane opened the pull request:
Closes https://github.com/rubocop-hq/rubocop/issues/7737

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

bbatsov commented:
I think that's not the best approach to building such a cop. A much better one would be to avoid hardcoding anything and build the rules for the methods and their redundant params via the configuration of the cop.

bbatsov commented on lib/rubocop/cop/style/unnecessary_arguments.rb:
```diff
@@ -0,0 +1,42 @@
+      #
+      # @example
+      #   # bad
+      #
```
You misplaced the blank line here.

bbatsov commented on lib/rubocop/cop/style/unnecessary_arguments.rb:
```diff
@@ -0,0 +1,42 @@
+module RuboCop
+  module Cop
+    module Style
+      # Checks for unnecessary arguments to builtin methods
```
I'd expand on the description here (and ideally go beyond builtin methods with some form of configuration).

koic commented on lib/rubocop/cop/style/unnecessary_arguments.rb:
```diff
@@ -0,0 +1,42 @@
+      #   string.split
+      #   "first second".split
+      #
+      class UnnecessaryArguments < Cop
```
For example, `RedundantSeparatorArgument` cop may be preferred for cop naming.
https://github.com/rubocop-hq/rubocop/issues/7077

tejasbubane replied:
Since @bbatsov suggested extending this cop beyond builtin methods, naming this `RedundantArguments`.

koic commented on config/default.yml:
```diff
@@ -3898,6 +3898,12 @@ Style/UnlessElse:
 
+Style/UnnecessaryArguments:
+  Description: 'Check for unnecessary arguments to builtin methods.'
+  Enabled: true
```
Can you specify `Enabled: pending` instead of `Enabled: true`.

tejasbubane replied:
Changed to `pending`.

tejasbubane commented on lib/rubocop/cop/style/redundant_arguments.rb:
```diff
@@ -0,0 +1,73 @@
+          when "''", '""'
+            ''
+          when '" "', "' '"
+            ' '
```
This is a bad workaround & will fail with cases like `'    '` (string with spaces). Can anyone help me fix this?

bbatsov replied:
I guess first you can squeeze the extra spaces before passing them to the cases. Maybe someone else from @rubocop-hq/rubocop-core has better ideas.

bquorning replied:
You could convert the configured values into AST nodes and compare directly against `node.arguments.first`.

I played a bit with it and got as far as this:
```rb
require 'parser/current'

def on_send(node)
  return unless redundant_argument?(node)
  # ...
end

def redundant_argument?(node)
  redundant_argument = redundant_arg_for_method(node.method_name.to_s)
  return false if redundant_argument.nil?

  node.arguments.first == redundant_argument
end

def redundant_arg_for_method(method_name)
  return nil unless cop_config['RedundantArguments'].key?(method_name)

  @mem ||= {}
  @mem[method_name] ||= begin
    arg = cop_config['RedundantArguments'].fetch(method_name)

    buffer = Parser::Source::Buffer.new('(string)', 1)
    buffer.source = arg.inspect
    builder = RuboCop::AST::Builder.new
    Parser::CurrentRuby.new(builder).parse(buffer)
  end
end
```

I am not _completely_ sure that a new buffer and builder instance is needed for each configuration, so the memoization may need to be moved around a bit.

tejasbubane commented:
@bbatsov Changed cop logic to be configurable for any method & argument.

bbatsov commented:
@tejasbubane Looks good to me overall. I'd only expand the docs to mention that people should be aware that the cop can't tell apart methods with the same name in different classes, so they should be careful with the cop.

bbatsov commented:
And that this is limited to methods that take a single param. To make it more generic we should be passing both a param position and a param in the config (e.g. if the redundant param is the second one), but I guess such cases are quite uncommon.

bbatsov commented on lib/rubocop/cop/style/redundant_arguments.rb:
```diff
@@ -0,0 +1,64 @@
+        MSG = 'Use `%<without_argument>s` instead of `%<with_argument>s`.'
+
+        def on_send(node)
+          arg = cop_config['RedundantArguments'][node.method_name.to_s]
```
Might be a good idea to add some validation logic for the config, so we can present the users with more meaningful error messages if they mess something up.

bbatsov commented on lib/rubocop/cop/style/redundant_arguments.rb:
```diff
@@ -0,0 +1,64 @@
+      #   "first second".split
+      #   A.foo
+      class RedundantArguments < Cop
+        MSG = 'Use `%<without_argument>s` instead of `%<with_argument>s`.'
```
I guess it'd be clearer to say that the arg is redundant.

bbatsov commented:
@tejasbubane ping :-)

tejasbubane commented:
@bbatsov Changes done.

marcandre commented:
Technically this cop is unsafe, as the actual default argument to join is `$OUTPUT_FIELD_SEPARATOR || ''` and that of split is `$FIELD_SEPARATOR || ' '`; if these globals are set, then `join()` and `split()` have different meaning than `join('')` and `split(' ')`

bbatsov commented:
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

tejasbubane commented:
@bbatsov Looks like I don't have the rights to reopen this PR. Do you mind if I open a new one?

tejasbubane commented:
> Technically this cop is unsafe, as the actual default argument to join is $OUTPUT_FIELD_SEPARATOR || '' and that of split is $FIELD_SEPARATOR || ' '; if these globals are set, then join() and split() have different meaning than join('') and split(' ')

@marcandre Thanks. The cop is marked unsafe. Also, these kernel variables are [deprecated since 2.7](https://github.com/ruby/ruby/pull/2920)

marcandre commented:
> I don't have the rights to reopen this PR. Do you mind if I open a new one?

Can't either. Says the branch was force-pushed or recreated. Not sure why github prevents it. Anyways, you can open a new one.

>  The cop is marked unsafe. 

Good, I must have forgotten to check, sorry.
